jacobian_follower/src/jacobian_follower/jacobian_follower.cpp

```diff
@@ -175,105 +141,123 @@ PlanResult JacobianFollower::plan(planning_scene_monitor::LockedPlanningSceneRW
-                                                                start_state,
-                                                                target_point_sequence);
+    auto const [traj, gripper_reached_target] = moveInWorldFrame(
+        planning_scene, group_name, tool_names, preferred_tool_orientations_q, start_state, target_point_sequence);
     reached_target = reached_target and gripper_reached_target;
 
     robot_trajectory.append(traj, 0);
   }
 
-  if (!time_param_.computeTimeStamps(robot_trajectory, max_velocity_scaling_factor, max_acceleration_scaling_factor))
-  {
+  if (!time_param_.computeTimeStamps(robot_trajectory, max_velocity_scaling_factor, max_acceleration_scaling_factor)) {
     ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "Time parametrization for the solution path failed.");
   }
 
-  if (not robot_trajectory.empty())
-  {
+  if (not robot_trajectory.empty()) {
     visual_tools_.publishTrajectoryPath(robot_trajectory);
   }
 
   return {robot_trajectory, reached_target};
 }
 
-bool JacobianFollower::isRequestValid(std::string const &group_name,
-                                      std::vector<std::string> const &tool_names,
-                                      std::vector<std::vector<Eigen::Vector3d>> const &grippers) const
-{
-  if (not model_->hasJointModelGroup(group_name))
-  {
+std::vector<std::vector<double>> JacobianFollower::compute_IK_solutions(geometry_msgs::Pose target_pose,
+                                                                        const std::string &group_name) {
+  auto const jmg = model_->getJointModelGroup(group_name);
+  const kinematics::KinematicsBaseConstPtr &solver = jmg->getSolverInstance();
+
+  Eigen::Isometry3d solverTrobot = Eigen::Isometry3d::Identity();
+
+  robot_state::RobotStatePtr kinematic_state = std::make_shared<robot_state::RobotState>(model_);
+  kinematic_state->setToIKSolverFrame(solverTrobot, solver);
+
+  // Convert to solver frame
+  Eigen::Affine3d ee_target_pose;
+  tf::poseMsgToEigen(target_pose, ee_target_pose);
+
+  Eigen::Affine3d pt_solver = solverTrobot * ee_target_pose;
+
+  std::vector<geometry_msgs::Pose> target_poses;
+  geometry_msgs::Pose pose;
+  tf::poseEigenToMsg(pt_solver, pose);
+  target_poses.push_back(pose);
+
+  //TODO: Add seed as optional input
+  std::vector<double> seed(jmg->getActiveJointModelNames().size());
+  std::vector<std::vector<double>> solutions;
+  kinematics::KinematicsResult result{};
+  kinematics::KinematicsQueryOptions options;
+  options.discretization_method = kinematics::DiscretizationMethod::ALL_DISCRETIZED;
+
+  solver->getPositionIK(target_poses, seed, solutions, result, options);
+  return solutions;
+}
+
+geometry_msgs::Pose JacobianFollower::computeFK(const std::vector<double> &joint_angles,
+                                                const std::string &group_name) {
+  auto jmg = model_->getJointModelGroup(group_name);
+  robot_state::RobotStatePtr kinematic_state(new robot_state::RobotState(model_));
```

Comment:
prefer make shared, it's shorter and less error prone
```suggestion
  auto kinematic_state = std::make_shared<robot_state::RobotState>(model_);
```